Add documention for the ACLK (#8413)
netdata/netdata: aclk/README.md

date: 2020-03-23
custom_edit_url: https://github.com/netdata/netdata/edit/master/aclk/README.md
---
-->

# Agent-cloud link (ACLK)

The agent-cloud link (ACLK) is the mechanism responsible for connecting a Netdata agent to Netdata Cloud. The ACLK uses
[MQTT](https://en.wikipedia.org/wiki/MQTT) over secure websockets to create, persist, and encrypt the connection from
end-to-end, and enable the features found in Netdata Cloud.

The ACLK is a vital component of the entirely redesigned and re-architected Netdata Cloud.

_No data is exchanged with Netdata Cloud until you claim a node._ By claiming a node, you opt-in to sending data from
your agent to Netdata Cloud via the ACLK. Data used to render charts and alarms in Netdata Cloud is accessed directly
from agents through the ACLK. Though the data does flow through Netdata servers on its way from agents to the browser,
it is never logged or stored.

Read our [claiming documentation](../claim/README.md) for details on node claiming.

## Enable and configure the ACLK

The ACLK is enabled by default if its prerequisites installed correctly, and the correct configuration is already set in
your agent's `netdata.conf` file.

```conf
[cloud]

>     cloud base url = https://up.netdata.cloud

Reviewer: app.netdata.cloud

```

### Proxy configuration

If your Netdata agent uses a proxy to reach the outside internet, you must configure a SOCKS5 proxy in the
`[agent_cloud_link]` section of your `netdata.conf` file. By default, the section looks like the following:

```conf
[agent_cloud_link]
    proxy = env
```

The `proxy` setting takes a few different values:

> -   `env`: Netdata reads the environment variables `http_proxy` and `socks_proxy` to discover the correct

Reviewer 2: maybe we can add that if both are defined `socks_proxy` is preferred. If this is not satisfactory and it is not desired to modify environment variables setting fixed proxy will work (as per info 2 lines below).

Author: Updated this section with some more information. Let me know if it looks OK.

The `proxy` setting takes a few different values:

-   `env`: Netdata reads the environment variables `socks_proxy` and `http_proxy` to discover the correct
    proxy settings.
-   `none`: Do not use any proxy, even if system is configured otherwise.
-   `socks5[h]://[user:pass@]host:ip`: Netdata uses the specified SOCKS proxy.
-   `http://[user:pass@]host:ip`: Netdata uses the specified HTTP proxy.

## Disable the ACLK

If you would prefer to disable the ACLK and not use Netdata Cloud, you have two options:

1.  Pass `--disable-cloud` to `netdata-installer.sh` during installation. When you pass this parameter, the installer
    does not download or compile any extra libraries, and the agent behaves as though the ACLK, and thus Netdata Cloud,
    does not exist.

2.  Change a runtime setting in your `netdata.conf` file. This setting only stops the agent from attempting any
    connection via the ACLK, but does not prevent the installer from downloading and compiling the ACLK's dependencies.

> 🎆 Needed: Configuration flag for disabling ACLK.

```conf
[global]
    ???
```

> 

Reviewer: Add instruction to remove the /etc/netdata/claim directory (unclaim an agent)

Author: Wouldn't this be better suited for the claiming README?

<!--
---
title: "Agent-cloud link (ACLK)"
description: "The agent-cloud link (ACLK) is the mechanism responsible for connecting a Netdata agent to Netdata Cloud. 
The ACLK uses MQTT over secure websockets to create, persist, and encrypt the connect from end-to-end, and enable the 
features found in Netdata Cloud."
date: 2020-03-16
custom_edit_url: https://github.com/netdata/netdata/edit/master/aclk/README.md
---
-->

# Agent-cloud link (ACLK)

The agent-cloud link (ACLK) is the mechanism responsible for connecting a Netdata agent to Netdata Cloud. The ACLK uses
[MQTT](https://en.wikipedia.org/wiki/MQTT) over secure websockets to create, persist, and encrypt the connect from
end-to-end, and enable the features found in Netdata Cloud.

The ACLK is a vital component of the entirely redesigned and re-architected Netdata Cloud.

_No data is exchanged with Netdata Cloud until you claim a node._ By claiming a node, you opt-in to sending data from
your agent to Netdata Cloud via the ACLK.

> ## Claim nodes with the ACLK

Reviewer 3: This overlaps with claim/README.md - should we merge docs?

Author: Good call. I'm going to put some of this over in `claim/README.md` and just link from one doc to the other.

> To use the ACLK to add a node to Netdata Cloud, you must **claim** the node.

Reviewer 4: ```suggestion
When you use the ACLK to add a node to Netdata Cloud, you must **claim** the node.
```

Claiming nodes is a security feature. Through the process of claiming, you demonstrate in a few ways that you have
administrative access to that node and the configuration settings for its Netdata agent. By logging into the node, you
prove you have access, and by running the `netdata-cloud.sh` script, you prove you have write access and administrative
privileges.

The claiming process ensures the ACLK will never be used by a third party to add your node, and thus view your metrics,
via their Netdata Cloud account.

> When you first sign in to the new Netdata Cloud, it will help onboard you to the new platform by asking you to **claim**

Reviewer 4: ```suggestion
When you first sign in to the new Netdata Cloud, it helps onboard you to the new platform by asking you to **claim**
```

any nodes you want to add to Netdata Cloud. The claiming process involves running a command, provided to you by the

> onboarding process, on each node to establish the token and use the ACLK to create a connection between the node and

Reviewer 4: ```suggestion
onboarding process, on each node to establish the token, and use the ACLK to create a connection between the node and
```

Netdata Cloud.

```bash

> netdata-cloud.sh --token 5AJRdC9H.XytDT8yWZB5psoy.BzCQz5IRm3uqo.nHNl9HQG3nEiTbPBxu_qYW2xdeFsEqFboD2.sf6zVPuveRtK3Xc1ouuDOS9hiCCLJRx1IBvktsvD16zH57tfCYh3kWxjEQH4 --rooms general,web

Reviewer 3: `netdata-claim.sh`

```

> The `netdata-cloud.sh` command takes two arguments:

Reviewer 3: `netdata-claim.sh`

-   `--token`: The unique token to your Netdata Cloud account to establish the ACLK.

> -   `--rooms`: A comma-separated list of the War Rooms this node should belong to.

Reviewer 3: We've added a `-proxy` option as well, details are in `claim/README.md`

> > It may take up to 60 seconds for your node to connect to Netdata Cloud after executing the `netdata-cloud.sh` command.

Reviewer 3: Global search/replace :)

> When a user of a Space visits an agent claimed by that Space, they automatically grant that agent access to their Cloud

Reviewer 3: This doesn't sound right - could somebody on the cloud team check this? From our perspective (on the agent) the visit to an already claimed node would not have any effect.

Author: This is from netdata/product#285, a specific request from @rduffield. Maybe it's not valid any more, but that's why it's there.

data via the ACLK.

## Enable and configure the ACLK

The ACLK is enabled by default if its prerequisites installed correctly, and the correct configuration is already set in
your agent's `netdata.conf` file.

> > 🎆 Needed: Any other configuration settings?

Reviewer 3: We just started discussing this, the current state of the code expects a `[agent_cloud_link]` section but this does not match the section naming conventions and there is no point separating the options. How would it look if we changed the layout of the config to this (@cosmix please check) :
```
[cloud]
  cloud base url = blahbhlahs
  proxy = socks5h://fhdjkfhjd
  server cert = ...
```
I don't like the repetition of `cloud` in both the section name and the setting but it is exposed via the registry / API and would break compatibility if we change it.

```conf
[cloud]
    cloud base url = https://up.netdata.cloud
```

### Proxy configuration

If your Netdata agent uses a proxy to reach the outside internet, you must configure a SOCKS5 proxy using one of the
following methods.

> 🎆 Needed: Configuration options/env variables for proxy.

## Disable the ACLK

> If you would prefer to disable the ACLK and not use Netdata Cloud, you have two options.

Reviewer 4: ```suggestion
If you would prefer to disable the ACLK and not use Netdata Cloud, you have two options:
```

> The first option is to pass `--disable-cloud` to `netdata-installer.sh` during installation. When you pass this

Reviewer 4: ```suggestion
1.   Pass `--disable-cloud` to `netdata-installer.sh` during installation. When you pass this
```

parameter, the installer does not download or compile any extra libraries, and the agent behaves as though the ACLK, and
thus Netdata Cloud, does not exist.

> The second option is to change a runtime setting in your `netdata.conf` file. This setting will only stop the agent from

Reviewer 4: ```suggestion
2.   Change a runtime setting in your `netdata.conf` file. This setting only stops the agent from
```

> attempting any connection via the ACLK, but will not prevent the installer from downloading and compiling the ACLK's

Reviewer 4: ```suggestion
attempting any connection via the ACLK, but does not prevent the installer from downloading and compiling the ACLK's
```

installation base. The ACLK should work on every Netdata agent, but there may be edge cases of incompatibility our team
is not yet aware of.

If the ACLK fails to build during manual installation, you may see one of the following error messages:

-   Failed to build libmosquitto. The install process will continue, but you will not be able to connect this node to
    Netdata Cloud.
-   Unable to fetch sources for libmosquitto. The install process will continue, but you will not be able to connect
    this node to Netdata Cloud.
-   Failed to build libwebsockets. The install process will continue, but you may not be able to connect this node to
    Netdata Cloud.
-   Unable to fetch sources for libwebsockets. The install process will continue, but you may not be able to connect
    this node to Netdata Cloud.

If your Netdata agent updates automatically, you can also look for error messages in `/var/log/netdata/error.log`. Try
one of the following two commands to search for ACLK-related errors.

```bash
less /var/log/netdata/error.log
grep ACLK /var/log/netdata/error.log
```

> To ACLK troubleshooting help from our engineers, [create an issue on

Reviewer 3: To -> For?

<!--
---
title: "Agent-cloud link (ACLK)"
description: "The agent-cloud link (ACLK) is the mechanism responsible for connecting a Netdata agent to Netdata Cloud. 
The ACLK uses MQTT over secure websockets to create, persist, and encrypt the connect from end-to-end, and enable the 
features found in Netdata Cloud."
date: 2020-03-18
custom_edit_url: https://github.com/netdata/netdata/edit/master/aclk/README.md
---
-->

# Agent-cloud link (ACLK)

The agent-cloud link (ACLK) is the mechanism responsible for connecting a Netdata agent to Netdata Cloud. The ACLK uses

> [MQTT](https://en.wikipedia.org/wiki/MQTT) over secure websockets to create, persist, and encrypt the connect from

Reviewer 5: s/connect/connection

end-to-end, and enable the features found in Netdata Cloud.

The ACLK is a vital component of the entirely redesigned and re-architected Netdata Cloud.

> _No data is exchanged with Netdata Cloud until you claim a node._ By claiming a node, you opt-in to sending data from

Reviewer 5: I like the word "exchanged". Is it worth explicitly mentioning that we do not store metric data?

Author: I think that's a good idea. Could you write up something along the lines of what you're thinking? It might make sense to even explain how metrics are passed via the ACLK to a user's browser, but I'm not aware enough of how it works at this moment to accurately explain it myself.

Reviewer 5: I’ll draft something as a comment here.

Reviewer 5: Maybe something like this?

> Data used to render charts and alarms in Netdata Cloud is accessed directly from agents through the ACLK. Though the data does flow through Netdata servers on its way from agents to the browser, it is never logged or stored.

Author: I like it. Added!

your agent to Netdata Cloud via the ACLK. Read our [claiming documentation](../claim/README.md) for details on node
claiming.

## Enable and configure the ACLK

The ACLK is enabled by default if its prerequisites installed correctly, and the correct configuration is already set in
your agent's `netdata.conf` file.

> 🎆 Needed: Any other configuration settings?

```conf
[cloud]
    cloud base url = https://up.netdata.cloud
```

### Proxy configuration

If your Netdata agent uses a proxy to reach the outside internet, you must configure a SOCKS5 proxy using one of the
following methods.

> > 🎆 Needed: Configuration options/env variables for proxy.

Reviewer 5: I put mine near the top of `/usr/sbin/netdata-claim.sh`, and it works:

```
ryan@ryan-VirtualBox:~$ head /usr/sbin/netdata-claim.sh
#!/usr/bin/env bash
# netdata
# real-time performance and health monitoring, done right!
# (C) 2017 Costa Tsaousis <costa@tsaousis.gr>
# SPDX-License-Identifier: GPL-3.0-or-later

export https_proxy=socks5://127.0.0.1:11081

# Exit code: 0 - Success
# Exit code: 1 - Unknown argument
```

Reviewer 5: I haven't tried with any kind of authentication, though.

Reviewer 3: You can do that with the -proxy switch instead (its in the README.md in the claim directory) or by exporting https_proxy in the shell (so that curl picks it up).

### Proxy configuration

If your Netdata agent uses a proxy to reach the outside internet, you must configure a SOCKS5 or HTTP proxy in the
`[cloud]` section of your `netdata.conf` file. A SOCKS5 proxy is preferred. By default, the section looks
like the following:

```conf
[cloud]
    proxy = none
```

The `proxy` setting takes a few different values:

-   `none`: Do not use any proxy, even if the system is configured otherwise.
-   `env`: Netdata reads the environment variables `socks_proxy` and `http_proxy` to discover the correct
    proxy settings.
-   `socks5[h]://[user:pass@]host:ip`: Netdata uses the specified SOCKS proxy.
-   `http://[user:pass@]host:ip`: Netdata uses the specified HTTP proxy.

## Disable the ACLK

If you would prefer to disable the ACLK and not use Netdata Cloud, you have two options:

1.  Pass `--disable-cloud` to `netdata-installer.sh` during installation. When you pass this parameter, the installer
    does not download or compile any extra libraries, and the agent behaves as though the ACLK, and thus Netdata Cloud,
    does not exist.

2.  Change a runtime setting in your `netdata.conf` file. This setting only stops the agent from attempting any

>     connection via the ACLK, but does not prevent the installer from downloading and compiling the ACLK's dependencies.

Reviewer 2: Maybe would add a sentence at the end?:
"This means ACLK functionality is available in the agent but remains fully inactive."

Reviewer 2: Additionally by default this key is set to `coming soon` in agent. This means now  same as `disable` and after the cloud release it will mean same as `enable`. Should we inform user about that? (to not make them confused when they see it in config?)

Reviewer 2: Hmmm, OK I guess then people who have config with `coming soon` in it from previous versions will be a minority or they will just rewrite it to `enable`/`disable` upon noticing it.

```

## Troubleshooting

The ACLK is a new and complex feature, but we're committed to maintaining compatibility across all of Netdata's
installation base. The ACLK should work on every Netdata agent, but there may be edge cases of incompatibility our team
is not yet aware of.

If the ACLK fails to build during manual installation, you may see one of the following error messages:

-   Failed to build libmosquitto. The install process will continue, but you will not be able to connect this node to
    Netdata Cloud.
-   Unable to fetch sources for libmosquitto. The install process will continue, but you will not be able to connect
    this node to Netdata Cloud.
-   Failed to build libwebsockets. The install process will continue, but you may not be able to connect this node to
    Netdata Cloud.
-   Unable to fetch sources for libwebsockets. The install process will continue, but you may not be able to connect
    this node to Netdata Cloud.

If your Netdata agent updates automatically, you can also look for error messages in `/var/log/netdata/error.log`. Try
one of the following two commands to search for ACLK-related errors.

```bash
less /var/log/netdata/error.log

> grep ACLK /var/log/netdata/error.log

Reviewer 2: `grep -i ACLK /var/log/netdata/error.log` just to be paranoid?

from agents through the ACLK. Though the data does flow through Netdata servers on its way from agents to the browser,
it is never logged or stored.

Read our [claiming documentation](../claim/README.md) for details on node claiming.

## Enable and configure the ACLK

The ACLK is enabled by default if its prerequisites installed correctly, and the correct configuration is already set in
your agent's `netdata.conf` file.

```conf
[cloud]
    cloud base url = https://app.netdata.cloud
```

Unless your node uses a proxy to connect with the internet, you do not need to configure the ACLK further. See the next
step for proxy configuration.

### Proxy configuration

If your Netdata agent uses a proxy to reach the outside internet, you must configure a SOCKS5 or HTTP proxy in the
`[agent_cloud_link]` section of your `netdata.conf` file. A SOCKS5 proxy is preferred. By default, the section looks
like the following:

```conf
[agent_cloud_link]
    proxy = env
```

The `proxy` setting takes a few different values:

-   `env`: Netdata reads the environment variables `socks_proxy` and `http_proxy` to discover the correct
    proxy settings.

> -   `none`: Do not use any proxy, even if system is configured otherwise.

Reviewer 3: ```suggestion
-   `none`: Do not use any proxy, even if the system is configured otherwise.
```

-   `socks5[h]://[user:pass@]host:ip`: Netdata uses the specified SOCKS proxy.
-   `http://[user:pass@]host:ip`: Netdata uses the specified HTTP proxy.

## Disable the ACLK

If you would prefer to disable the ACLK and not use Netdata Cloud, you have two options:

1.  Pass `--disable-cloud` to `netdata-installer.sh` during installation. When you pass this parameter, the installer
    does not download or compile any extra libraries, and the agent behaves as though the ACLK, and thus Netdata Cloud,
    does not exist.

2.  Change a runtime setting in your `netdata.conf` file. This setting only stops the agent from attempting any
    connection via the ACLK, but does not prevent the installer from downloading and compiling the ACLK's dependencies.

> 🎆 Needed: Configuration flag for disabling ACLK.

```conf
[global]

>     ???

Reviewer 3: We haven't decided the name of the flag yet, but I was thinking:
```
[global]
    netdata cloud = coming soon            # Default for now
    netdata cloud = enable                      # Default after release
    netdata cloud = disable                     # How to switch everything off
```
@cosmix / @cakrit - do you have a preference here?
